clang/lib/Sema/SemaDecl.cpp

```diff
@@ -12053,6 +12053,10 @@ bool Sema::CheckFunctionDeclaration(Scope *S, FunctionDecl *NewFD,
   if (LangOpts.OpenMP)
     OpenMP().ActOnFinishedFunctionDefinitionInOpenMPAssumeScope(NewFD);
 
+  if (LangOpts.isSYCL() && NewFD->hasAttr<SYCLKernelEntryPointAttr>() &&
+      !NewFD->isInvalidDecl() && !NewFD->isDependentContext())
```

Comment:
Hmm, I'm not a fan of interfaces that implicitly do nothing under some set of conditions; I prefer that callers be required to consider whether those conditions require some other action.

Reply:
I've been thinking the opposite lately, which is why I made this comment.  These conditions end up getting 'forgotten' somewhere along the way, and we get something that acts weird in release build.

Reply:
That is exactly my concern and why I want such mistakes to be caught at compile-time. This is the motivation for ensuring that an incorrect registration fails compilation even if a check or diagnostic is missed somewhere.

Reply:
At this point though, you're not actually catching anything at compile-time.  All you're doing is asserting at runtime if someone forgets one of these tags, rather than just preventing us from having to worry about it.

Reply:
The call to `llvm::report_fatal_error()` will occur at run-time for Clang, sure, but that will be compile-time for users building their SYCL source code. I would rather have Clang fail with a stack trace that clearly indicates the problem (to us) then perform code gen thinking that a kernel was registered when one wasn't (likely to manifest in some kind of "kernel not found" error at run-time) or, worse, having a duplicate registration silently rejected such that an unintended kernel gets called. I grant that these scenarios will only occur if there are missing or incorrect checks elsewhere, but these kinds of problems would be rather painful to trace back to what the problem really is.

Reply:
And I'm saying that is SUCH a divergence from the standard behavior of the compiler and our current coding standards, that I don't want to do it the first time here.

We likely want some sort of RFC for that which audits all current asserts/unreachables/etc, and figures out which of those should be `report_fatal_error`, and a comprehensive patch to fix all/most of them at once.

Reply:
This is not the first such case. A few existing examples are listed below. More examples can be found in Serialization.
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/AST/ExternalASTSource.cpp#L121
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/ProfileList.cpp#L62
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/SanitizerSpecialCaseList.cpp#L36
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Basic/SourceManager.cpp#L674
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/CoverageMappingGen.cpp#L1030-L1031
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/CoverageMappingGen.cpp#L1050-L1051
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/LinkInModulesPass.cpp#L30
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/CodeGen/ObjectFilePCHContainerWriter.cpp#L277
- https://github.com/llvm/llvm-project/blob/b1320d36339e38b073088fd45013a3c692adb301/clang/lib/Interpreter/Interpreter.cpp#L426-L428

Reply:
I don't doubt that others got through review, and a few of those I clicked on are different enough situations as to not be the same as this. This is a novel use of that function, and one caught during review.

Reply:
I think this is substantially similar. It seems we aren't going to agree. Let's see what others have to say.

Reply:
I'm going to resolve this conversation about the use of `llvm::report_fatal_error()` since the most recent comments on this topic are happening in https://github.com/llvm/llvm-project/pull/111389#discussion_r1825122665.